test(client): custom output with yarn

millsp opened the pull request:
closes https://github.com/prisma/team-orm/issues/866
closes https://github.com/prisma/prisma/issues/17685

millsp committed:
fix missing version in custom output

millsp committed:
add test for yarn

millsp commented on packages/client/src/generation/generateClient.ts:
```diff
@@ -133,6 +133,7 @@ export async function buildClient({
     types: 'index.d.ts',
     browser: 'index-browser.js',
     exports: clientPackageExports,
+    version: clientVersion,
```
Found that this was missing when writing the test.

millsp replied:
I had to put the PR in draft again. Somehow changing either `version` or `private` causes our functional type tests to fail. I don't understand how they influence each other.

SevInf replied:
So, i figured that out and fixed. It's quite a wild thing: If we have multiple npm packages with the same name and version in the tree and they all contain the same namespace declaration, TS will merge those together into single namespace. Not having a `version` in generated `package.json` prevented that. Now, when we added the version, all `Prisma` namespace declaration from all of the tests got merged together, sometimes overriding each other's properties, sometimes producing intersections that result in `never`` and so on. 

I've fixed that in my commit here by appending sha256 of the schema file to a generated package name. We might also consider making generated package name configurable.

millsp committed:
exclude yarn lock

millsp committed:
comment version out

millsp committed:
remove private instead

millsp committed:
use current client version

Jolg42 commented on packages/client/tests/e2e/custom-output-yarn/readme.md:
```diff
@@ -0,0 +1,11 @@
+# Readme
+
+This tests ensures that custom output clients are working as expected with Yarn.
```
```suggestion
This tests ensures that custom output clients are working as expected with Yarn 1.
```
I'm not sure how this goes with other versions

SevInf replied:
Same with Berry

Jolg42 commented on packages/client/tests/e2e/custom-output-yarn/readme.md:
```diff
@@ -0,0 +1,11 @@
+
+This tests ensures that custom output clients are working as expected with Yarn.
+
+A package is deemed not to be valid by yarn if:
```
```suggestion
A `package.json` is invalid for `yarn` if:
```

SevInf committed:
Ensure unique generated package name

That prevents TS from merging `Prisma` namespace declarations from
different clients together.

SevInf committed:
Update packages/client/tests/e2e/custom-output-yarn/readme.md

Co-authored-by: Joël Galeran <Jolg42@users.noreply.github.com>
